libraries/chain/finality/qc.cpp

```diff
@@ -309,10 +365,55 @@ std::optional<qc_t> aggregating_qc_t::get_best_qc(block_num_type block_num) cons
    return std::optional<qc_t>{qc_t{block_num, std::move(*active_best_qc), {}}};
 }
 
+// A dual finalizer votes on both active and pending finalizer policies.
+void aggregating_qc_t::verify_dual_finalizers_votes(const qc_t& qc) const {
+   auto build_vote_index = [&](const std::vector<finalizer_authority>& finalizers,
+                               std::map<bls_public_key, uint32_t>& index) {
+      uint32_t i = 0;
+      for (const auto& f: finalizers) {
+         index[f.public_key] = i;
+         ++i;
+      }
+   };
+
+   std::map<bls_public_key, uint32_t> active_finalizer_vote_index; // from public key to vote bit index
+   build_vote_index(active_finalizer_policy->finalizers, active_finalizer_vote_index);
+
+   std::map<bls_public_key, uint32_t> pending_finalizer_vote_index; // from public key to vote bit index
+   build_vote_index(pending_finalizer_policy->finalizers, pending_finalizer_vote_index);
+
+   std::vector<bls_public_key> pending_finalizer_keys_sorted; // all pending finalizer public keys sorted
+   pending_finalizer_keys_sorted.reserve(pending_finalizer_policy->finalizers.size());
+   for (const auto& f: pending_finalizer_policy->finalizers) {
+      pending_finalizer_keys_sorted.emplace_back(f.public_key);
+   }
+   std::sort(pending_finalizer_keys_sorted.begin(), pending_finalizer_keys_sorted.end());
+
+   // Find dual finalizers (which vote on both active and pending policies)
+   // and verify each dual finalizer votes in the same way.
+   for (const auto& f: active_finalizer_policy->finalizers) {
+      bool found = std::binary_search(pending_finalizer_keys_sorted.begin(),
+                                      pending_finalizer_keys_sorted.end(),
+                                      f.public_key);
```

Comment:
Instead of sorting (which requires copying all the `bls_public_key`s), I think it would be better to just sequential scan the entire vector each time. `bls_public_key` is rather large. I wonder if we would be better off using `shared_ptr<bls_public_key>` everywhere instead.

Reply:
It is now much simpler. Thanks, https://github.com/AntelopeIO/spring/pull/475/commits/f1f3ac2befc80c8ffd3faa0916970f792b09b4ec